propose chain_index::Queryable as the name of the capability

zancas opened the pull request:
Fixes: #719 

Renaming the ChainIndex to Queryable has a few benefits.

(1) It hints to a reader that the thing is not a type.
(2) It conforms to a notion of a Capability that a thing can be made to offer by implementing it.
(3) It frees the name "ChainIndex" to be used less ambiguously on related types.
(4) It reduces a redundancy in the naming scheme when the mod name is also accounted for.

At least as a thought experiment, @nachog00 reconsidered related trait and type names and scopes and opened this PR as a proposal for an update to the naming scheme.

zancas committed:
propose chain_index::Query as the name of the capability

zancas committed:
NBCI --> ChainIndex, and move to top of file

zancas committed:
NodeBackedChainIndexSubscriber --> Subscriber

zancas committed:
propose that nonfinalized is inside snapshot

zancas committed:
add file

zancas committed:
Client --> Index

zancas committed:
fix comments

zancas committed:
ChainIndex --> Query

zancas committed:
chain_index --> index

AloeareV commented on zaino-state/src/chain_index.rs:
```diff
@@ -162,10 +295,10 @@ mod tests;
-/// 3. Call `NodeBackedChainIndex::new(source, config).await`
-pub trait ChainIndex {
+/// 3. Call `Index::new(source, config).await`
+pub trait Query {
```
A query is, intuitively, the thing that you pass as an argument to this type's methods and get back a response. Calling this thing itself a `Query` is misleading.

zancas replied:
But..  a trait is not a "thing".

zancas replied:
How about "Queryable"?

AloeareV replied:
Much clearer!

AloeareV commented on zaino-state/src/chain_index.rs:
```diff
@@ -419,23 +419,23 @@ impl<Source: BlockchainSource> NodeBackedChainIndex<Source> {
     }
 
-    /// Creates a [`NodeBackedChainIndexSubscriber`] from self,
+    /// Creates a [`Subscriber`] from self,
```
A lot of information has been removed from this type name. Why? This name is now less informative to me.

zancas replied:
The "ChainIndex" component (especially to internal developers) is read from the mod name.

Per the `NodeBacked` qualifier..  are there other implementors/providers?

nachog00 replied:
even if there aren't any other ones, the main one that is aimed to be the one consumed in production is, to me, worthy of the context-less name

zancas replied:
@nachog00 and I suspected that _NodeBacked_ might have been useful to discriminate from the trait named **ChainIndex**, but if we now name that trait **Queryable** then there's one less reason to prepend **NodeBacked**.

AloeareV commented on zaino-state/src/chain_index/snapshot.rs:
```diff
@@ -0,0 +1,61 @@
+
+use std::sync::Arc;
+
+impl<T> NonFinalized for Arc<T>
```
I think if we're to remove words from this trait name (again, less information? Why?) then it's more useful to call this thing a 'Snapshot'. The fact that it is of the non-finalized state is useful for understanding the underlying mechanics of the snapshot process, but it's irrelevant to the consumer.

AloeareV replied:
(whoops. This comment is on the impl block, meant to put it on the NonFinalized trait itself)

nachog00 replied:
Since this is the snapshot module, and it provides the impl to get a snapshot of the `non_finalised_state`, it seems to me like this trait should be called something along the lines of `Snapshottable` or plain `Snapshot`.

zancas replied:
> again, less information? Why?

I think less information in the last part of the name, can stimulate the reader to consider the other parts of the name..  i.e. the mod path.    So..   often I think it's better to not replicate a mod name in the items it contains..   this practice can couple nicely with a practice of partial path imports..   so one could:

```Rust
use chain_index::snapshot;

let x = snapshot::NonFinalized
```

In that way the explicit and verbose naming information IS included but with the context that let's the reader immediately know about the definition's local mod hierarchy.

AloeareV commented on zaino-state/src/chain_index/non_finalised_state.rs:
Discussed on call...The non-finalized state is used by the chain index a lot. If there's a module/submodule relation, than snapshot should be inside non_finalized_state and not the other way around...however, the snapshot being of the non-finalized is useful information for a zaino-developer...but at the public API level this is irrelevant, the snapshot is a component of the chain index. I think the `snapshot` and `non_finalized_state` mods should both be in the `chain_index` mod.

zancas replied:
You prefer `snapshot` and `non_finalized_state` to be sibling mods within `chain_index`?

nachog00 replied:
Is it right to say that the chain index takes a snapshot of itself?

AloeareV replied:
I do. Another possibility that I just thought of is to have the snapshot mod be inside the non_finalized_state mod, make sure that it's not public, and then pub use the snapshot mod in chain_index. This means that snapshot is clearly inside non_finalized_state from a developer perspective, but make them siblings at the public API level (as the snapshot being of the nonfinalized state is useful for developers to know, but unnecessary confusion to consumers)

zancas replied:
If this were the organization then the name of the trait might make more sense as "SnapShotable"?  Instead of "NonFinalized"?

AloeareV commented on zaino-state/src/lib.rs:
```diff
@@ -31,13 +31,14 @@ pub use backends::{
 
 // Core ChainIndex trait and implementations
-pub use chain_index::{ChainIndex, NodeBackedChainIndex, NodeBackedChainIndexSubscriber};
+pub use chain_index::Query as ChainIndexQuery;
```
I said something about this in a previous comment, but to elaborate:

The trait `Read` means 'something that is readable'. `Read` as a trait is intuitively read as something readable...as there's no noun 'read' that you use in order to read (verb). 

Query, however, *is* a noun, meaning something like 'the request you make of a database in order to learn information from it', so when I see something called a `Query` I don't automatically think about it as something that can be queried...rather, the more obvious way to read this is that the thing called `Query` is, in fact, a `Query`, rather than 'something that is queryable'.

zancas replied:
Heh...  great minds think alike.

zancas committed:
flip such that non_finalised_state contains snapshot

zancas committed:
Query --> Queryable

zancas committed:
begin defining snapshot and non_finalised_state as sibling mods within chain_index

zancas committed:
make snapshot and non_finalised_state sibling mods within chain_index

zancas committed:
clarify doc link

zancas commented:
I fixed a doc lint, and I think I addressed all of the current comments.   I didn't resolve any conversations, as I suppose it *might* make more sense for the initiator of conversations to click resolve.

I'm hoping that @AloeareV and/or @nachog00 are interested in continuing the conversation that this PR is derived from.

I'd be curious if @nachog00 had any comments to add here.

zancas committed:
Merge branch 'dev' into propose_chain_index_query_name

nachog00 commented on zaino-state/src/chain_index.rs:
```diff
@@ -97,13 +231,13 @@ mod tests;
 /// let snapshot = subscriber.snapshot_nonfinalized_state();
 ///
-/// // Query blocks in a range using the subscriber
+/// // Queryable blocks in a range using the subscriber
```
Query seemed right on this comment before the change

nachog00 commented on zaino-state/src/chain_index.rs:
```diff
@@ -132,7 +266,7 @@ mod tests;
 /// ).await?;
 ///
-/// // Wrap the connector for use with ChainIndex
+/// // Wrap the connector for use with Queryable
```
fully qualified `chain_index::Queryable` would make more sense on this doc line.

nachog00 commented on zaino-state/src/chain_index.rs:
````diff
@@ -44,23 +45,156 @@ pub mod types;
 /// ```no_run
 /// # async fn example() -> Result<(), Box<dyn std::error::Error>> {
-/// use zaino_state::{ChainIndex, NodeBackedChainIndex, ValidatorConnector, BlockCacheConfig};
+/// use zaino_state::{Queryable, Index, ValidatorConnector, BlockCacheConfig};
````
this import should be through the `chain_index` module path, for clarity

nachog00 commented on zaino-state/src/chain_index.rs:
```diff
@@ -97,13 +231,13 @@ mod tests;
+/// let subscriber = index.subscriber().await;
 ///
 /// // Take a snapshot for consistent queries
 /// let snapshot = subscriber.snapshot_nonfinalized_state();
```
is `.snapshot_non_finalized_state()` a good method name?
if we are providing it on the `Index` type, then should it not be `.snapshot()` and have consumers abstracted from the details?

AloeareV replied:
Good call

zancas committed:
update to latest dev and resolve conflict

zancas committed:
NonFinalized --> Snapshottable
